From 64739ec2e2b9b6025c99329fdf0738b142ce8d9c Mon Sep 17 00:00:00 2001
From: Kevin Hao <kexin.hao@windriver.com>
Date: Fri, 19 Jul 2019 13:31:21 +0800
Subject: [PATCH] octeontx2-pf: Disable the BH when invoking
 otx2_napi_handler() in non softirq context

The function otx2_napi_handler() is supposed to be run under the
softirq context, so we should disable the BH if we invoke it under
non softirq context. Otherwise we would get a call trace like this:
    BUG: using __this_cpu_add() in preemptible [00000000] code: ethtool/583
    caller is __this_cpu_preempt_check+0x18/0x20
    CPU: 15 PID: 583 Comm: ethtool Not tainted 4.18.40-yocto-standard+ #35
    Hardware name: Marvell OcteonTX CN96XX board (DT)
    Call trace:
     dump_backtrace+0x0/0x168
     show_stack+0x24/0x30
     dump_stack+0x80/0xa4
     check_preemption_disabled+0xfc/0x100
     __this_cpu_preempt_check+0x18/0x20
     __netif_receive_skb_core+0xdc/0xb98
     __netif_receive_skb+0x34/0x88
     netif_receive_skb_internal+0x70/0x168
     napi_gro_receive+0xc0/0x188
     otx2_napi_handler+0x4f0/0x870
     otx2_free_hw_resources+0x3d8/0x698
     otx2_stop+0x18c/0x228
     otx2_set_ringparam+0xf0/0x150
     dev_ethtool+0xd48/0x2310
     dev_ioctl+0x19c/0x3e8
     sock_ioctl+0x278/0x4d8
     do_vfs_ioctl+0xc4/0x878
     ksys_ioctl+0x8c/0xa0
     sys_ioctl+0x34/0x48
     el0_svc_naked+0x30/0x34

Signed-off-by: Kevin Hao <kexin.hao@windriver.com>
---
 drivers/net/ethernet/marvell/octeontx2/nic/otx2_pf.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/drivers/net/ethernet/marvell/octeontx2/nic/otx2_pf.c b/drivers/net/ethernet/marvell/octeontx2/nic/otx2_pf.c
index 6e90bcf892c2..e2398f0da2cf 100644
--- a/drivers/net/ethernet/marvell/octeontx2/nic/otx2_pf.c
+++ b/drivers/net/ethernet/marvell/octeontx2/nic/otx2_pf.c
@@ -1462,8 +1462,11 @@ static void otx2_free_hw_resources(struct otx2_nic *pf)
 		cq = &qset->cq[qidx];
 		cqe_count = otx2_read64(pf, NIX_LF_CINTX_CNT(cq->cint_idx));
 		cqe_count &= 0xFFFFFFFF;
-		if (cqe_count)
+		if (cqe_count) {
+			local_bh_disable();
 			otx2_napi_handler(cq, pf, cqe_count);
+			local_bh_enable();
+		}
 	}
 
 	/* Free RQ buffer pointers*/
-- 
2.17.1

